xapian-letor/scorer/err_score.cc

```diff
@@ -0,0 +1,96 @@
+/** @file err_score.cc
+ *  @brief Implementation of ERRScore
+ *
+ *  ERR Score is adapted from the paper: http://olivier.chapelle.cc/pub/err.pdf
+ *  Chapelle, Metzler, Zhang, Grinspan (2009)
+ *  Expected Reciprocal Rank for Graded Relevance
+ */
+/* Copyright (C) 2014 Hanxiao Sun
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/scorer.h"
+
+#include "debuglog.h"
+
+#include <algorithm>
+
+using namespace std;
+
+using namespace Xapian;
+
+ERRScore::ERRScore()
+{
+    LOGCALL_CTOR(API, "ERRScore", NO_ARGS);
+}
+
+ERRScore::~ERRScore()
+{
+    LOGCALL_DTOR(API, "ERRScore");
+}
+
+double
+ERRScore::score(const std::vector<FeatureVector> & fvv) const
+{
+    LOGCALL(API, double, "ERRScore::score", fvv);
+    if (fvv.empty()) {
+	throw Xapian::InvalidArgumentError("Empty argument supplied");
+    }
+    int length = fvv.size();
+
+    /* used to store values which change from labels to
+     * satisfaction probability
+     */
+    double intermediate_values[length];
```

Comment:
This uses a non-standard extension supported by GCC and some other compilers - ISO C++ requires the size of an array to be known at compile time.  We should avoid non-standard extensions (except in code which is inherently compiler-specific anyway).

(I know James said to use an array instead, but I think he overlooked that the array size here isn't a compile-time constant).

The options I can see are:

(a) Revert to using `vector<double>`.

(b) Use `double* intermediate_values = new double[length];` - the downside is you then need to explicitly call `delete []` on the returned pointer, and that's fiddly to ensure in the face of exceptions.

(c) Use `std::unique_ptr<double[]> intermediate_values{new double[length]};` - effectively (b) but using `unique_ptr` to ensure there's no leak.

(d) Do away with `intermediate_values` entirely.  Calling `fvv[i].get_label()` returns a member variable rather than performing some complex calculation, so there's not really a good reason to carefully cache the result - we could just call this again in the final loop.

If we need a temporary container, (c) is the cleanest and least overhead, but I'd suggest (d) makes more sense in this case.

Reply:
yes, definitely d makes more sense